atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCellsReader.java

```diff
@@ -0,0 +1,139 @@
+    }
+
+    private void populateResults(SweepableCellsTable.SweepableCellsRow row, Map.Entry<Cell, Value> entry,
+            Map<TableReferenceAndCell, Long> results) {
+        SweepableCellsTable.SweepableCellsColumn col = SweepableCellsTable.SweepableCellsColumn.BYTES_HYDRATOR
+                .hydrateFromBytes(entry.getKey().getColumnName());
+
+        if (isReferenceToDedicatedRows(col)) {
+            populateFromDedicatedRows(row, col, results);
+        } else {
+            populateFromValue(row, col, entry.getValue(), results);
+        }
+    }
+
+    private boolean isReferenceToDedicatedRows(SweepableCellsTable.SweepableCellsColumn col) {
+        return col.getWriteIndex() < 0;
+    }
+
+    private void populateFromDedicatedRows(SweepableCellsTable.SweepableCellsRow row,
+            SweepableCellsTable.SweepableCellsColumn col, Map<TableReferenceAndCell, Long> results) {
+        TargetedSweepMetadata metadata = TargetedSweepMetadata.BYTES_HYDRATOR.hydrateFromBytes(row.getMetadata());
+        long partitionFine = row.getTimestampPartition();
+        int numberOfDedicatedRows = (int) -col.getWriteIndex();
+
+        List<byte[]> dedicatedRows = new ArrayList<>();
+        for (int i = 0; i < numberOfDedicatedRows; i++) {
+            byte[] metadataBytes = ImmutableTargetedSweepMetadata.builder()
+                    .from(metadata)
+                    .dedicatedRow(true)
+                    .dedicatedRowNumber(i)
+                    .build()
+                    .persistToBytes();
+            dedicatedRows.add(SweepableCellsTable.SweepableCellsRow.of(partitionFine, metadataBytes).persistToBytes());
+        }
+        RowColumnRangeIterator iterator = getAllColumns(dedicatedRows);
+        iterator.forEachRemaining(entry -> populateResults(row, entry, results));
+    }
+
+    private void populateFromValue(SweepableCellsTable.SweepableCellsRow row,
+            SweepableCellsTable.SweepableCellsColumn col,
+            Value value,
+            Map<TableReferenceAndCell, Long> results) {
+        TableReferenceAndCell tableRefCell = SweepableCellsTable.SweepableCellsColumnValue
+                .hydrateValue(value.getContents());
+
+        long timestamp = row.getTimestampPartition() * SweepQueueUtils.TS_FINE_GRANULARITY + col.getTimestampModulus();
+        addIfMaxForCell(timestamp, tableRefCell, results);
+    }
+
+    private void addIfMaxForCell(long ts, TableReferenceAndCell refAndCell, Map<TableReferenceAndCell, Long> result) {
+        result.merge(refAndCell, ts, Math::max);
+    }
+
+    private static ColumnRangeSelection allPossibleColumns() {
+        byte[] startCol = SweepableCellsTable.SweepableCellsColumn.of(0L, -TargetedSweepMetadata.MAX_DEDICATED_ROWS)
+                .persistToBytes();
+        byte[] endCol = SweepableCellsTable.SweepableCellsColumn.of(SweepQueueUtils.TS_FINE_GRANULARITY + 1, 0)
+                .persistToBytes();
+        return new ColumnRangeSelection(startCol, endCol);
```

Comment:
I think this covers all the columns we need to. The first component ranges from `0` to `TS_FINE_GRANULARITY - 1` (both inclusive); the second component ranges from `-MAX_DEDICATED_ROWS` to `MAX_DEDICATED_ROWS` (both inclusive).

There are tighter bounds for endCol (I think you can actually specify `SweepQueueUtils.TS_FINE_GRANULARITY - 1, TargetedSweepMetadata.MAX_DEDICATED_ROWS + 1`) though they shouldn't affect anything. I'd actually suggest taking `TS_FINE_GRANULARITY` straight (i.e. without the `+ 1`) for readability.

Reply:
I think this version now is the easiest to read

Reply:
Yep, this is what I would prefer as well. Thanks